src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/TextBasedLogicalPlanPrinterImpl2.java

```diff
@@ -0,0 +1,314 @@
+package se.liu.ida.hefquin.engine.queryplan.utils;
+
+import java.io.PrintStream;
+
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlanVisitor;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlanWalker;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.*;
+import se.liu.ida.hefquin.engine.utils.IndentingPrintStream;
+
+public class TextBasedLogicalPlanPrinterImpl2 implements LogicalPlanPrinter
+{
+	// The string represents '|  '.
+	protected final String levelIndentBase = "\u2502  ";
+	// The string represents '├──'.
+	protected final String nonLastChildIndentBase = "\u251C\u2500\u2500";
+	// The string represents '└──'.
+	protected final String lastChildIndentBase = "\u2514\u2500\u2500";
+		
+	@Override
+	public void print( final LogicalPlan plan, final PrintStream out ) {
+		final IndentingPrintStream iOut = new IndentingPrintStream(out);
+		//final LogicalPlanVisitor beforeVisitor = new MyBeforeVisitor(iOut);
+		//final LogicalPlanVisitor afterVisitor = new MyAfterVisitor(iOut);
+		//LogicalPlanWalker.walk(plan, beforeVisitor, afterVisitor);
+		//iOut.flush();
+		
+		planWalk(plan, 0, 0, 0, iOut);
+		iOut.flush();
+		
+	}
+	
+	public String getIndentLevelString(int planNumber, int planLevel, int numberOfSiblings) {
+		String indentLevelString = "";
+		for ( int i = 1; i < planLevel; i++ ) {
+			indentLevelString += levelIndentBase;
+		}
+		if (planNumber < numberOfSiblings-1) {
+			indentLevelString += nonLastChildIndentBase ;
+		}
+		else {
+			if (numberOfSiblings > 0) {
+				indentLevelString += lastChildIndentBase;
+			}
+			else {
+				indentLevelString = "";
+			}
+		}
+		return indentLevelString;
+	}
+	
+	public void planWalk( final LogicalPlan plan, int planNumber, int planLevel, int numberOfSiblings, IndentingPrintStream out) {
+		//System.out.println(getIndentLevelString2(planLevel, planNumber, numberOfSiblings) + plan.getRootOperator().toString() +" "+planLevel + " " + planNumber + " "+ numberOfSiblings);
+		//System.out.println(getIndentLevelString2(planLevel, planNumber, numberOfSiblings) + plan.getRootOperator().toString() );
+		//System.out.println(plan.getRootOperator().printString(getIndentLevelString2(planLevel, planNumber, numberOfSiblings)));
+		out.append( plan.getRootOperator().printString(getIndentLevelString(planNumber, planLevel, numberOfSiblings)) );
+		out.append(System.lineSeparator());
```

Comment:
Be consistent in terms of how you use spaces within such function calls. Otherwise the code looks messy. For instance, in the first of these two lines, you have a space before the first argument and a space after the last argument, whereas the second line does not have such spaces.

My rule of thumb for these cases: If all the arguments are only program variables (i.e., no function calls themselves and no use of the dot (`.`) operator), then I omit these spaces. As soon as one of the arguments is not just a program variable, I put these spaces (and then always both at the beginning and at the end, as you did in the first of these two lines here). Hence, the first of these two lines satisfies this rule but the second one does not.

Also, try to avoid nesting of function calls to a level of two or more. That makes the code difficult to read, and step-by-step debugging also becomes more difficult to keep track of where the execution currently is.